### ISSUE TITLE
Refresh not working with auth code flow?

### ISSUE THREAD

**Issue by fragsalat** [spotify/spotify-web-api-ts-sdk#79]
Hey there,

I want to use the library in a nodejs service where access and refresh token are already prepared. Thus I use the `withAccessToken()` method. When calling the `authenticate()` function with an expired token it obviously tries to refresh it. Sadly the client receives `invalid_request` from the api. I suspect the reason is the missing Authorization header. The API specifies the refresh request to use basic authentication with client_id and secret_key and supply `grant_type=refresh_token&refresh_token=...`. Additionally a client_id body param is required for PKCE method. From the code https://github.com/spotify/spotify-web-api-ts-sdk/blob/main/src/auth/AccessTokenHelpers.ts#L28 we see there is not authorization header passed. 

Am I doing or use the library in a wrong way? Or is that a missing feature?

Example code
```
  const spotify = SpotifyApi.withAccessToken(config.spotify.clientId, {
    token_type: 'Bearer',
    access_token: config.spotify.accessToken,
    refresh_token: config.spotify.refreshToken,
    expires_in: config.spotify.expiresAt.getTime() - Date.now(),
    expires: config.spotify.expiresAt.getTime(),
  });

  const params = request.nextUrl.searchParams;
  const search = params.get('search');
  const type = params.get('type');

  // Will attempt refresh of token and fails with invalid_request
  await spotify.authenticate()

  const result = await spotify.search(search, [type as ItemTypes])
```

Thanks in advance and best regards

**Comment by fragsalat** [comment]
I also checked if it would work to pass a custom fetch function, where I patch the headers object with my authorization header. But sadly the fetch of the SdkOptions isn't used by the AccessTokenHelper, which is used to refresh the token.

**Comment by zhammer** [comment]
Hitting the same issue. The issue is that there are two auth flows involving [refresh tokens](https://developer.spotify.com/documentation/web-api/tutorials/refreshing-tokens): the [Authorization Code Flow](https://developer.spotify.com/documentation/web-api/tutorials/code-flow) and the [Authorization Code with PKCE Flow](https://developer.spotify.com/documentation/web-api/tutorials/code-pkce-flow). The two have different security use cases.

A major difference between the two (besides slightly different request & response types) is that in PKCE, when a token is refreshed, it returns back in its response both an `access_token` and a `refresh_token`. In PKCE, `refresh_token`s can be used only once, which will grant a new, one-time-only refresh token and an `access_token`.

In non-PKCE, when a `refresh_token` is refreshed, only a new access token is returned, and the old `refresh_token` can be used again with no expiry.

The default [`refreshTokenAction`](https://github.com/spotify/spotify-web-api-ts-sdk/blob/5d14751c6881d7831f60c1e785cb43697df2658f/src/auth/ProvidedAccessTokenStrategy.ts#L21) on the `ProvidedAccessTokenStrategy` assumes the provided `refresh_token` comes from the PKCE flow and uses that flow in refreshing that token.

If `ProvidedAccessTokenStrategy` were exported, like @fragsalat suggests, we could pass in a non-PKCE function like the following:

```ts
async function refreshNonPKCEAccessToken(clientId: string, item: AccessToken): Promise<ICachable & AccessToken> {
  const params = new URLSearchParams();
  params.append("grant_type", "refresh_token")
  params.append("refresh_token", MY_REFRESH_TOKEN)

  const result = await fetch('https://accounts.spotify.com/api/token', {
    method: 'POST',
    headers: {
      'Authorization': `Basic ${Buffer.from(`${clientId}:${MY_CLIENT_SECRET}`).toString('base64')}`,
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body: params,
  })
  const text = await result.text();
  if (!result.ok) {
    throw new Error(`Failed to refresh token: ${result.statusText}, ${text}`);
  }
  const json = JSON.parse(text);
  return json;
}
```

**Comment by EduardsE** [comment]
@zhammer Can you show an example of how to use this fix? I'm having the same issue, also using `withAccessToken` auth method.

**Comment by EduardsE** [comment]
@JustinApt tried the fix but still getting that same `Bad or expired token`. I'm probably doing something wrong.

**Comment by EduardsE** [comment]
This is how my code looks now and `refreshNonPKCEAccessToken` never seems to get called.

```typescript
async function refreshNonPKCEAccessToken(clientId: string, item: AccessToken): Promise<ICachable & AccessToken> {
  const params = new URLSearchParams();
  params.append("grant_type", "refresh_token");
  params.append("refresh_token", item.refresh_token);

  const result = await fetch("https://accounts.spotify.com/api/token", {
    method: "POST",
    headers: {
      Authorization: `Basic ${Buffer.from(`${clientId}:${process.env!.SPOTIFY_CLIENT_SECRET}`).toString("base64")}`,
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: params,
  });
  const text = await result.text();
  if (!result.ok) {
    throw new Error(`Failed to refresh token: ${result.statusText}, ${text}`);
  }
  const json = JSON.parse(text);
  return json;
}

function SpotifyApiConstructor(clientId: string, token: AccessToken, config?: SdkOptions) {
  const strategy = new ProvidedAccessTokenStrategy(clientId, token, () => refreshNonPKCEAccessToken(clientId, token));
  return new SpotifyApi(strategy, config);  
}

class Spotify {
  private api: SpotifyApi;

  constructor(private user: User) {
    this.api = SpotifyApiConstructor(process.env!.SPOTIFY_CLIENT_ID!, {
      access_token: user.accessToken,
      refresh_token: user.refreshToken,
      expires_in: user.expiresIn,
      token_type: "Bearer",
      expires: Date.now() + user.expiresIn,
    });
  }

  public getPlaylist = (playlistId: string) => {
    return this.api.playlists.getPlaylist(playlistId);
  };
}

```

**Comment by EduardsE** [comment]
The fix works, I simply had wrong `expiresIn` and `expires` values so it never tried to refresh the token. I somehow was under the impression that if API returns `Bad or expired token` error, it would refresh token and try to call endpoint again but that's not how that works.

**Comment by EduardsE** [comment]
@JustinApt  `expires_in` should be in seconds. When a user authenticates, that access token is viable for 1 hour our 3600 seconds. So if user authenticated half an our ago, in `expires_in` you should put 1800.
And expires should contain the time the access token expires at. So if user authenticated half an our ago, `expires` should be `Date.now() + 1800`.

This is how it looks for me. `user.expires` is the a date that comes from database
```typescript    
this.api = SpotifyApiConstructor(process.env!.SPOTIFY_CLIENT_ID!, {
      access_token: user.accessToken,
      refresh_token: user.refreshToken,
      token_type: "Bearer",
      expires: Number(user.expires),
      expires_in: Math.round((new Date(user.expires).getTime() - new Date().getTime()) / 1000),
    });
```

**Comment by fragsalat** [comment]
Thanks for the fix and sorry for coming back that late. The adjustment is what I was looking for. Thanks for that. I think from my side this can be closed :)

**Comment by fragsalat** [comment]
One last thing, could  @fostertheweb maybe create a PR to the original repository to get a new version published? So far it is only available in the fork or?

**Comment by fostertheweb** [comment]
@fragsalat unfortunately, there hasn't been any activity here for 3+ months. I've merged many of the open PRs for this repo into my fork and have published them to npm for us all to use.

**Comment by danielvy** [comment]
I am using a local build with these changes applied but it would be nice to have them finally published.


And just a couple of things worth mentioning that may save others looking for this solution some time:

Calling the mentioned `refreshNonPKCEAccessToken` with a refresh token obtained from client PKCE authorization will not convert the refresh token to a long term one. It will immediately invalidate the previous refresh token and return a new single-use expiring one. So you can't just reuse the token obtained from the client with PKCE authorization flow.

In order to have a long term refresh token you need your user to explicitly go through non PKCE [authorization flow](https://developer.spotify.com/documentation/web-api/tutorials/code-flow) first to obtain the token (server side, don't expose your client secret to the user). Then you can keep using the same refresh token with `refreshNonPKCEAccessToken`.

If you have the long term refresh token, you don't have to store or update anything else in your permanent storage. You can always use this refresh token to obtain a fresh access token at any time until the refresh token expires/invalidated. Just check that the refresh token hasn't been changed (which can happen under certain circumstances).

Using an access token obtained from long term refresh token with the default `ProvidedAccessTokenStrategy` will invalidate the refresh token when the access token expires. So don't mix the two.
